pallets/gear-messenger/src/migrations.rs

```diff
@@ -0,0 +1,435 @@
+// This file is part of Gear.
+
+// Copyright (C) 2023 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{Config, DispatchStash, Dispatches, Pallet, Waitlist};
+use common::storage::{Interval, LinkedNode};
+use frame_support::{
+    traits::{Get, GetStorageVersion, OnRuntimeUpgrade},
+    weights::Weight,
+};
+use frame_system::pallet_prelude::BlockNumberFor;
+use gear_core::ids::MessageId;
+use parity_scale_codec::Encode;
+use sp_std::marker::PhantomData;
+#[cfg(feature = "try-runtime")]
+use {
+    frame_support::{codec::Decode, dispatch::DispatchError},
+    sp_std::vec::Vec,
+};
+
+pub struct MigrateToV3<T: Config>(PhantomData<T>);
+
+impl<T: Config> OnRuntimeUpgrade for MigrateToV3<T> {
+    fn on_runtime_upgrade() -> Weight {
+        let current = Pallet::<T>::current_storage_version();
+        let onchain = Pallet::<T>::on_chain_storage_version();
+
+        log::info!(
+            "🚚 Running migration with current storage version {current:?} / onchain {onchain:?}"
+        );
+
+        // 1 read for on chain storage version
+        let mut weight = T::DbWeight::get().reads(1);
+
+        if current == 3 && onchain == 2 {
+            Waitlist::<T>::translate(
+                |_, _, store: (v2::StoredDispatch, Interval<BlockNumberFor<T>>)| {
+                    weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+                    Some((store.0.into(), store.1))
+                },
+            );
+
+            Dispatches::<T>::translate(|_, store: LinkedNode<MessageId, v2::StoredDispatch>| {
+                weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+                Some(LinkedNode {
+                    next: store.next,
+                    value: store.value.into(),
+                })
+            });
+
+            DispatchStash::<T>::translate(
+                |_, store: (v2::StoredDispatch, Interval<BlockNumberFor<T>>)| {
+                    if store.0.context.is_some() {
```

Comment:
could you add similar check in pre_uprade, that none of delayed dispatches has context?